Integrate swap_tensors into nn.Module.load_state_dict

mikaylagawarecki opened the pull request:
Added a `torch.Tensor` method that defines how to transform `other`, a value in the state dictionary, to be loaded into `self`, a param/buffer in an `nn.Module` before swapping via `torch.utils.swap_tensors`
* `param.module_load(sd[key])`

This method can be overridden using `__torch_function__`.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117913

mikaylagawarecki committed:
[WIP] Integrate swap_tensors into nn.Module.load_state_dict

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Integrate swap_tensors into nn.Module.load_state_dict"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Integrate swap_tensors into nn.Module.load_state_dict"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Integrate swap_tensors into nn.Module.load_state_dict"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Integrate swap_tensors into nn.Module.load_state_dict"

Added two `torch.Tenosr` methods `dest.module_load_from(src)` and `src.module_load_to(dest)` that define how to transform `src` when loading `src` from a `state_dict` into `dest` before using `torch.utils.swap_tensors()` in `load_state_dict()`

These methods can be overridden using `__torch_function__`, if so they are and torch.utils.get_swap_module_params_on_conversion` is `True`, then `swap_tensors` will be used in `m.load_state_dict()`





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Integrate swap_tensors into nn.Module.load_state_dict"

Added two `torch.Tensor` methods that define how to transform `src`, a value in the state dictionary, to be loaded into `dest`, a param/buffer in an `nn.Module` before swapping via `torch.utils.swap_tensors`
* `dest.module_load_from(src)`
* `src.module_load_to(dest)` 

These methods can be overridden using `__torch_function__`, if they are and `torch.utils.get_swap_module_params_on_conversion` is `True`, then `swap_tensors` will be used in `m.load_state_dict()`





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Integrate swap_tensors into nn.Module.load_state_dict"

Added two `torch.Tensor` methods that define how to transform `src`, a value in the state dictionary, to be loaded into `dest`, a param/buffer in an `nn.Module` before swapping via `torch.utils.swap_tensors`
* `dest.module_load_from(src)`
* `src.module_load_to(dest)` 

These methods can be overridden using `__torch_function__`, if they are and `torch.utils.get_swap_module_params_on_conversion` is `True`, then `swap_tensors` will be used in `m.load_state_dict()`





[ghstack-poisoned]

albanD commented on test/nn/test_load_state_dict.py:
```diff
@@ -189,7 +191,8 @@ def my_state_dict_pre_hook(*args, **kwargs):
 
-    # FIXME: doesn't fail locally, maybe remove
+    # fails in crossref mode when swapping tensors as LSTM installs weakrefs
+    @skipIfCrossRef
```
Is that independant?

albanD commented on test/nn/test_load_state_dict.py:
```diff
@@ -362,7 +365,162 @@ def test_load_state_dict_warn_assign(self):
 
+class MyLoadTensor(torch.Tensor):
+
+    # enable deepcopy
```
deepcopy doesn't work without that?

mikaylagawarecki replied:
I think  `new_empty` needs to be implemented as `__deepcopy__` goes through this path for non-wrapper

https://github.com/pytorch/pytorch/blob/9fe3693bbb9494f57febebe0d6aefa9da081b16a/torch/_tensor.py#L172-L188

The override seemed to error if within the `__torch_function__` but that might just be a bug on my end, will retry.

albanD commented on test/nn/test_load_state_dict.py:
```diff
@@ -362,7 +365,162 @@ def test_load_state_dict_warn_assign(self):
+            elif type(src) is MyLoadTensor:
+                return src
+            elif type(src) is MyWrapperLoadTensor:
+                return MyLoadTensor(src._data)
```
What is `._data` ??

mikaylagawarecki replied:
`MyWrapperLoadTensor._data`, I named the attribute of the wrapper subclass as such, but can rename

albanD replied:
Ho my bad. I though it was a field on the base Tensor.

albanD commented on test/nn/test_load_state_dict.py:
```diff
@@ -362,7 +365,162 @@ def test_load_state_dict_warn_assign(self):
+        elif func is torch.Tensor.module_load_to:
+            return module_load_to(*args, **kwargs)
+        elif func is torch.Tensor.detach:
+            return MyLoadTensor(args[0].data.detach())
```
Don't use .data!
Also why is this needed?

albanD commented on test/nn/test_load_state_dict.py:
```diff
@@ -362,7 +365,162 @@ def test_load_state_dict_warn_assign(self):
+        return f"MyWrapperLoadTensor({self._data.__repr__()})"
+
+    # enable deepcopy
+    def new_empty(self, shape):
```
Similar: why?

mikaylagawarecki replied:
I think this was wrong, `new_empty` should only be needed for non-wrapper subclasses, my bad

albanD commented on test/nn/test_load_state_dict.py:
```diff
@@ -362,7 +365,162 @@ def test_load_state_dict_warn_assign(self):
+                return func(*args, **kwargs)
+
+
+class MyWrapperLoadTensor(torch.Tensor):
```
Inherit from MyLoadTensor to reduce duplication?

mikaylagawarecki replied:
Hmmm one is a wrapper subclass and one is not though, can I still inherit from the non-wrapper MyLoadTensor

albanD commented on test/nn/test_load_state_dict.py:
```diff
@@ -362,7 +365,162 @@ def test_load_state_dict_warn_assign(self):
+        def _test(module_subclass=None, sd_subclass=None):
+            m = _create_model(module_subclass)
+            sd = _create_model(sd_subclass).state_dict()
+            # deepcopy as swap_tensors might destructively modify state dict
```
Ho that is indeed an interesting side effect...

mikaylagawarecki replied:
The current state is that 
- swap_tensors does not destructively modify state dict
-  in the case where `param` or `sd[key]` has `__torch_function__` handler for `module_load` if `id(param.module_load(sd[key])) == id(sd[key])`, `param`'s `THPVariable->cdata` will point to the same `TensorImpl` as `sd[key]`'s `THPVariable->cdata`
- I added a warning in the `nn.Module.load_state_dict` docs rearding this 

After thinking about it more, I am not sure that we want to `clone()` the `sd[key]` to preserve behavior of the non-swap path. I think it is unlikely that people use the `state_dict` after `load_state_dict` but curious for your feedback on this!

albanD commented on test/nn/test_load_state_dict.py:
```diff
@@ -362,7 +365,162 @@ def test_load_state_dict_warn_assign(self):
+
+
+# run tests witsh swap enabled/disabled
+def instantiate_parametrize_with_swap(cls):
```
Why do you need this here? Can't it be in the utils or included in the parametrized test generation?

mikaylagawarecki replied:
Do you mean move this function to a test utils file?

I want to run each test within the parametrized class with two different decorators that contain context managers that set `swap`.

I did not want to include it in the parametrized test generation (`instantiate_parametrized_tests`) as that seemed too "large blast" and I don't think we need to parametrize that many tests with swap True and False

I think using `@parametrize` here would require explicitly using the context manager in every test within the test class which I preferred not to do :P

I agree that my solution here is not clean though, do you know of any utility that would achieve this?

mikaylagawarecki replied:
Oops ok changed this to a `@swap` `_TestParametrizer` and decorated each function

albanD commented on torch/_tensor.py:
```diff
@@ -710,6 +710,64 @@ def share_memory_(self):
+            other (Tensor): value in state dict with key corresponding to ``self``
+
+        """
+        if has_torch_function_unary(self):
```
You should be checking both self and other here. This is quite meh

mikaylagawarecki committed:
Update on "Integrate swap_tensors into nn.Module.load_state_dict"

Added two `torch.Tensor` methods that define how to transform `src`, a value in the state dictionary, to be loaded into `dest`, a param/buffer in an `nn.Module` before swapping via `torch.utils.swap_tensors`
* `dest.module_load_from(src)`
* `src.module_load_to(dest)` 

These methods can be overridden using `__torch_function__`, if they are and `torch.utils.get_swap_module_params_on_conversion` is `True`, then `swap_tensors` will be used in `m.load_state_dict()`





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Integrate swap_tensors into nn.Module.load_state_dict"

Added two `torch.Tensor` methods that define how to transform `src`, a value in the state dictionary, to be loaded into `dest`, a param/buffer in an `nn.Module` before swapping via `torch.utils.swap_tensors`
* `dest.module_load_from(src)`
* `src.module_load_to(dest)` 

These methods can be overridden using `__torch_function__`, if they are and `torch.utils.get_swap_module_params_on_conversion` is `True`, then `swap_tensors` will be used in `m.load_state_dict()`





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Integrate swap_tensors into nn.Module.load_state_dict"

Added two `torch.Tensor` methods that define how to transform `src`, a value in the state dictionary, to be loaded into `dest`, a param/buffer in an `nn.Module` before swapping via `torch.utils.swap_tensors`
* `dest.module_load_from(src)`
* `src.module_load_to(dest)` 

These methods can be overridden using `__torch_function__`, if they are and `torch.utils.get_swap_module_params_on_conversion` is `True`, then `swap_tensors` will be used in `m.load_state_dict()`





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Integrate swap_tensors into nn.Module.load_state_dict"

Added two `torch.Tensor` methods that define how to transform `src`, a value in the state dictionary, to be loaded into `dest`, a param/buffer in an `nn.Module` before swapping via `torch.utils.swap_tensors`
* `dest.module_load_from(src)`
* `src.module_load_to(dest)` 

These methods can be overridden using `__torch_function__`, if they are and `torch.utils.get_swap_module_params_on_conversion` is `True`, then `swap_tensors` will be used in `m.load_state_dict()`





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Integrate swap_tensors into nn.Module.load_state_dict"

Added two `torch.Tensor` methods that define how to transform `src`, a value in the state dictionary, to be loaded into `dest`, a param/buffer in an `nn.Module` before swapping via `torch.utils.swap_tensors`
* `dest.module_load_from(src)`
* `src.module_load_to(dest)` 

These methods can be overridden using `__torch_function__`, if they are and `torch.utils.get_swap_module_params_on_conversion` is `True`, then `swap_tensors` will be used in `m.load_state_dict()`





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Integrate swap_tensors into nn.Module.load_state_dict"

Added two `torch.Tensor` methods that define how to transform `src`, a value in the state dictionary, to be loaded into `dest`, a param/buffer in an `nn.Module` before swapping via `torch.utils.swap_tensors`
* `dest.module_load_from(src)`
* `src.module_load_to(dest)` 

These methods can be overridden using `__torch_function__`, if they are and `torch.utils.get_swap_module_params_on_conversion` is `True`, then `swap_tensors` will be used in `m.load_state_dict()`





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Integrate swap_tensors into nn.Module.load_state_dict"

Added a `torch.Tensor` methods that define how to transform `src`, a value in the state dictionary, to be loaded into `dest`, a param/buffer in an `nn.Module` before swapping via `torch.utils.swap_tensors`
* `param.module_load_from(sd[key])`

This method can be overridden using `__torch_function__`.




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Integrate swap_tensors into nn.Module.load_state_dict"

Added a `torch.Tensor` method that defines how to transform `src`, a value in the state dictionary, to be loaded into `dest`, a param/buffer in an `nn.Module` before swapping via `torch.utils.swap_tensors`
* `param.module_load(sd[key])`

This method can be overridden using `__torch_function__`.




[ghstack-poisoned]

albanD commented on test/nn/test_load_state_dict.py:
```diff
@@ -8,17 +10,29 @@
     import numpy as np
 
 
+class swap(_TestParametrizer):
```
Why isn't this living in common_utils with wrapSwapTensorsTest?

albanD commented on test/nn/test_load_state_dict.py:
```diff
@@ -362,7 +386,123 @@ def test_load_state_dict_warn_assign(self):
+        return module_load(*args, **kwargs)
+    else:
+        with torch._C.DisableTorchFunctionSubclass():
+            # detach must return instance of same subclass for nn.Parameter()
```
Is MyWrapperLoadTensor actually using this function though?

mikaylagawarecki replied:
My understanding is when creating `Parameter(MyWrapperLoadTensor)`

`MyWrapperLoadTensor.detach()` would call into `MyLoadTensor.__torch_function__` which would then go into the `__torch_dispatch__` for `MyWrapperLoadTensor.detach()`, so I do think it will go into this, yes.

albanD replied:
Ok!

mikaylagawarecki commented on torch/nn/modules/module.py:
```diff
@@ -2101,6 +2114,13 @@ def load_state_dict(self, state_dict: Mapping[str, Any],
             If :attr:`assign` is ``True`` the optimizer must be created after
             the call to :attr:`load_state_dict`.
 
+        .. warning::
```
This behavior differs from the default/might be surprising, so I thought it might be worth mentioning.

It is probably unlikely that the user will use the state_dict after loading though, so maybe it is okay to omit this

albanD replied:
The default is to .copy_() into what is here right now. So yes the new behavior differs from it but that is much broader than what you mention here.
Also I think the flag setter is the right place to explain what the flag does.

mikaylagawarecki committed:
Update on "Integrate swap_tensors into nn.Module.load_state_dict"

Added a `torch.Tensor` method that defines how to transform `src`, a value in the state dictionary, to be loaded into `dest`, a param/buffer in an `nn.Module` before swapping via `torch.utils.swap_tensors`
* `param.module_load(sd[key])`

This method can be overridden using `__torch_function__`.




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Integrate swap_tensors into nn.Module.load_state_dict"

Added a `torch.Tensor` method that defines how to transform `src`, a value in the state dictionary, to be loaded into `dest`, a param/buffer in an `nn.Module` before swapping via `torch.utils.swap_tensors`
* `param.module_load(sd[key])`

This method can be overridden using `__torch_function__`.




[ghstack-poisoned]

albanD commented on torch/__future__.py:
```diff
@@ -22,28 +25,41 @@ def set_overwrite_module_params_on_conversion(value: bool) -> None:
+
+    The semantics for :meth:`~nn.Module.load_state_dict` when this is set are as follows:
+
+    #. ``sd['key']`` is transformed via :meth:`~torch.Tensor.module_load`
```
Very clear explanation.
I don't think `sd['key']` is very meaningful though?

mikaylagawarecki replied:
```suggestion
    #. ``state_dict['key']`` is transformed via :meth:`~torch.Tensor.module_load`
```

albanD reviewed (approved):
Small nit but SGTM!

mikaylagawarecki commented on torch/__future__.py:
```diff
@@ -22,28 +25,41 @@ def set_overwrite_module_params_on_conversion(value: bool) -> None:
+
+    The semantics for :meth:`~nn.Module.load_state_dict` when this is set are as follows:
+
+    #. ``sd['key']`` is transformed via :meth:`~torch.Tensor.module_load`
+       (i.e. ``res = param.module_load(sd['key'])``)
```
```suggestion
    #. For each parameter/buffer, its corresponding``sd['key']`` is transformed via 
       :meth:`~torch.Tensor.module_load` (i.e. ``res = param.module_load(state_dict['key'])``)
```

mikaylagawarecki committed:
Update torch/__future__.py

mikaylagawarecki commented on torch/__future__.py:
```diff
@@ -22,28 +25,41 @@ def set_overwrite_module_params_on_conversion(value: bool) -> None:
+
+    The semantics for :meth:`~nn.Module.load_state_dict` when this is set are as follows:
+
+    #. For each parameter/buffer, its corresponding``sd['key']`` is transformed via 
```
```suggestion
    #. For each parameter/buffer, its corresponding``state_dict['key']`` is transformed via 
```

mikaylagawarecki committed:
Update torch/__future__.py

mikaylagawarecki committed:
Update on "Integrate swap_tensors into nn.Module.load_state_dict"

Added a `torch.Tensor` method that defines how to transform `src`, a value in the state dictionary, to be loaded into `dest`, a param/buffer in an `nn.Module` before swapping via `torch.utils.swap_tensors`
* `param.module_load(sd[key])`

This method can be overridden using `__torch_function__`.




[ghstack-poisoned]

mikaylagawarecki commented:
@pytorchbot merge
